### ISSUE TITLE
refactor usage of concurrency with ReactiveX

### ISSUE THREAD

**Issue by pollend** [MovingBlocks/Terasology#4798]
some situations the number of threads are overprovisioned and in some cases its under provisioned. Some workloads might benefit with more threads but selective changing the number of threads will be over provisioned on some systems and under provisioned on another system. 

### Current situation

- Chunk-Unloader (allocate 4 threads)
  - int list to queue event for BeforeDeactivateBlocks
  - very light load so threads are mostly idle
  -at any moment a single thread will become active
  
- Chunk-Updater (allocates 8 threads)
 - generates chunk mesh and queue them back on the main thread

- LOD Chunk Generation
 - only triggered for handling different levels of lod

- ThreadManagerSubsystem (allocates 16 threads)
 - loadingGame
 - create new world
 - saving screenshot

- TaskMaster
  - Behaviors (1 thread)
  - FlexiablePathFinding (4 threads)
  - Pathfinding (1 thread)

-----------------------------------

- [ ] drop usage of the ThreadManagerSubsystem easily can be replaced by usage of Observable
- [ ] remove usage of TaskMaster
  - [ ] replace usage of threading in Behaviours
  - [ ] replace usage of threading in FlexiablePathFinding
  - [ ] replace usage of threading in pathfinding
  - [ ] LOD Chunk generation
  - [ ] chunk unloader probably does not benefit with parallelism
- [ ] replace usage of threading in FlexiablePathFinding

**Comment by keturn** [comment]
Do we have anyone familiar enough with the various implementations of the reactive pattern in Java to evaluate [Project Reactor](https://projectreactor.io/) (with `Flux` and `Mono` APIs) in comparison to RxJava2?

**Comment by skaldarnar** [comment]
I tried to sketch out what we're currently doing and using (no guarantee for completeness):
![image](https://user-images.githubusercontent.com/1448874/123555764-0a1b0880-d788-11eb-914e-4b234cfde699.png)

**Comment by pollend** [comment]
@skaldarnar that's right got me confused because ExecutorService just creates a pool of threads and each instance of TaskMaster will introduce its own set of threads. we might get into a situation where the processor is constantly context switching between a lot of computational threads.

**Comment by pollend** [comment]
> Do we have anyone familiar enough with the various implementations of the reactive pattern in Java to evaluate [Project Reactor](https://projectreactor.io/) (with `Flux` and `Mono` APIs) in comparison to RxJava2?

there are a lot of weird facilities to map one to the other and vice versa. probably want to provide one use case so we don't end up juggling a lot of these reactive libraries.

**Comment by skaldarnar** [comment]
@pollend I've updated your initial post by adding links to ReactiveX concepts and tweaking the sentences a bit here and there. I hope I did not change the semantic anywhere.

> got me confused because ExecutorService just creates a pool of threads and each instance of TaskMaster will introduce its own set of threads

You plan forward sounds good to me on first read. I just wanted to share that picture I made, where we can hopefully remove a few of the green boxes in the future, and replace others with RX concepts.

**Comment by pollend** [comment]
https://github.com/MovingBlocks/Terasology/pull/4799 @skaldarnar this should correctly expose reactivex to module space. you need to use the GameThread wrapper though :/

**Comment by keturn** [comment]
I'd like the roadmap to include a plan for replacing the debug thread view; currently [ThreadMonitorPanel](https://github.com/MovingBlocks/Terasology/blob/0fa1272241b43be94e9920c84a41683643c9f240/engine/src/main/java/org/terasology/engine/monitoring/gui/ThreadMonitorPanel.java#L47) and the [RunningThreads metrics mode] use ThreadMonitor.

https://github.com/MovingBlocks/Terasology/issues/4637 links to a few options that work with ExecutorServices. At the time, I wasn't checking for integration with RxJava or Project Reactor. They might have something built-in, or a different integration option for these other libraries.

**Comment by pollend** [comment]
> I'd like the roadmap to include a plan for replacing the debug thread view; currently [ThreadMonitorPanel](https://github.com/MovingBlocks/Terasology/blob/0fa1272241b43be94e9920c84a41683643c9f240/engine/src/main/java/org/terasology/engine/monitoring/gui/ThreadMonitorPanel.java#L47) and the [RunningThreads metrics mode](https://github.com/MovingBlocks/Terasology/blob/0fa1272241b43be94e9920c84a41683643c9f240/engine/src/main/java/org/terasology/engine/rendering/nui/layers/ingame/metrics/RunningThreadsMode.java#L14) use ThreadMonitor.
> 
> #4637 links to a few options that work with ExecutorServices. At the time, I wasn't checking for integration with RxJava or Project Reactor. They might have something built-in, or a different integration option for these other libraries.

hmm, would be a nice way to do some kind of server metrics. processing results from the running tasks?

**Comment by pollend** [comment]
some main problems with RXJava. There is a table of mappings that define a flowable, Observable, Single, Maybe, and Completable and mappings from one to the other. 0..N defined for Floablw and Observable but only flowable has backpressure. Single is one element only but Maybe is 0 or 1. Completable is 0 element or error. there are different methods to map one type to another. the inconsistent API is difficult to follow. The mappings make this pretty confusing where reactor has Flux and Mono. Flux is 0...N and mono is just a single object. I like this a lot better then the former. 

While working with Flowable I managed to get into a situation where objects getting pushed to my Flowable were not getting handled by the subscriber where this pattern seems to be more reliable with reactor. I was looking at this benchmark comparing the two and found the reactor implementation is quite a bit faster: https://github.com/MovingBlocks/Terasology/pull/4786

https://github.com/akarnokd/akarnokd-misc/issues/7

**Comment by keturn** [comment]
Reactor explicitly has [built-in integration with micrometer](https://projectreactor.io/docs/core/release/reference/#metrics), which makes for a good default answer to which of the metrics implementations in #4637 we should use.

**Comment by pollend** [comment]
> Reactor explicitly has [built-in integration with micrometer](https://projectreactor.io/docs/core/release/reference/#metrics), which makes for a good default answer to which of the metrics implementations in #4637 we should use.

me and @DarkWeird were discussing this.

**Comment by skaldarnar** [comment]
We briefly discussed this in today's org meeting and I quickly updated the graphic to match the current state of things:

![untitled_page_1](https://user-images.githubusercontent.com/1448874/131224475-5877c71b-5b3a-47fe-9504-c6857aee860a.png)

**Comment by DarkWeird** [comment]
I  should refer my pr there  #4865

**Comment by DarkWeird** [comment]
And i have a several thinks about networking via reactor.
 Reactor project have `reactor-netty` lib:
1. network subsystem should looks smaller and easier with it.
2. We can use HttpClient from reactor-netty.
     It should make module downloader and ModuleListDownload more faster(?) And non blocking (especially for ModuleListDownloader, which use separate thread via new Thread :( )
3. We can make refactor of DiscordRPC. (-1 thread)

**Comment by jdrueckert** [comment]
So my first performance test plan was simply to spawn 42 deer in a CoreGameplay + WildAnimals module combo and look at the FPS...

Good news: With the current `FlexiblePathfinding` `develop`, I have 60 FPS
Bad news: With this branch checked out, I have 1.7 FPS - and no, the decimal point in there is no accident...
